### PR TITLE
Add egs_view egsinp editor with autocompletion for all egs++ apps

### PR DESCRIPTION
**Pull request by Xuan-Zheng05**
Add a new tab in egs_view that allows for editing and saving egsinp files for egs++.

The new code all over the egs++ library provide a structure by which an external library like egs_view can extract all of the input parameters. It's used for input validation, syntax highlighting, autocompletion, etc.

Recent changes include support for application inputs, run control, transport parameters. Pegless material support is in progress.

Note that this includes an astyle run on all the sources, geometries and ausgab objects, so it looks like there are more changes than there are. If you view the diff while hiding whitespace, then you will see the real changes.

This replaces PR https://github.com/nrc-cnrc/EGSnrc/pull/885.

### CODE REVIEW

**Comment by rtownson** [comment]
Rebased on develop.

This PR still needs a bit of love to finish off. Double checking the tooltips for consistency & usefulness, showing the 'current application' in the editor, and improving the names of the examples in the drop-down.

**Comment by mainegra** [comment]
@rtownson this branch still includes a bug introduced by commit a7ab9c37 into the egs_fluence_scoring AO when Grace file axis labels where being "fixed" for beamdp, but should not have been done for the C++ part of the codes/sources.
Funnily enough, you had fixed this on a later commit 1415fbb
Because of that, when I am configuring the system for testing this branch, the process crashes. :cry:

**Comment by rtownson** [comment]
@mainegra I think you're on an old version! Delete your local add_egs_view_editor3 and pull it from Xuan's repo again. There was a forced rebase to bring it up to date with `develop`.

I've tried to reduce the confusion by deleting the EGSnrc branch `add-egs-view-editor3`, which was created by mistake at one point. Now it should only be on Xuan's repo.

You can check if the latest commit is [d10d5f0](https://github.com/nrc-cnrc/EGSnrc/pull/1164/commits/d10d5f0740a35420ced980049b1751a25d5b57fb)

**Comment by mainegra** [comment]
> @mainegra I think you're on an old version! Delete your local add_egs_view_editor3 and pull it from Xuan's repo again. There was a forced rebase to bring it up to date with `develop`.
> 
> I've tried to reduce the confusion by deleting the EGSnrc branch `add-egs-view-editor3`, which was created by mistake at one point. Now it should only be on Xuan's repo.
> 
> You can check if the latest commit is [d10d5f0](https://github.com/nrc-cnrc/EGSnrc/pull/1164/commits/d10d5f0740a35420ced980049b1751a25d5b57fb)

@rtownson I simply looked at the branch name and checked it out locally. How do I pull from Xuan's repo?

**Comment by rtownson** [comment]
> @rtownson I simply looked at the branch name and checked it out locally. How do I pull from Xuan's repo?

To checkout an external PR:
```
git fetch origin pull/$ID/head:$BRANCHNAME
git checkout $BRANCHNAME
```

In this case:
```
git fetch origin pull/1164/head:add-egs-view-editor3
git checkout add-egs-view-editor3
```

**Comment by mainegra** [comment]
@rtownson now that I got the proper branch, I was able to configure the system without any problems!

**_I love the global opacity feature!_**

However, there were issues when attempting to load an `egs_chamber` input file: 
- it complained that `libegs_app.so` was not available:
```
❯ egs_view Dgas-NE2571-HDR.egsinp
EGS_Library::load(): failed to load library /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/libegs_app.so
  error was: /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/libegs_app.so: cannot open shared object file: No such file or directory

egspp: Failed to load the egs_app application library from /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/

malloc(): invalid size (unsorted)
Aborted (core dumped)
```
- After building `egs_app` then it finally loaded the geometry while issuing the message:
<br>

```
❯ egs_view Dgas-NE2571-HDR.egsinp

egspp: Failed to resolve the address of the 'createApplication' function in the application library /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/libegs_app.so
```
<br>

**Question:** Why ask for `libegs_app.so` if it can't load it?

Ah, reading will set me free! :-) One needs at least` libegs_app.so` to be able to use the viewer, right?

I also realized that I had to update both applications! Duh!

After doing so, and building both codes, I can't execute the viewer!

```
❯ egs_view Dgas-NE2571-HDR.egsinp

egspp: Failed to resolve the address of the 'createApplication' function in the application library /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/libegs_app.so

double free or corruption (!prev)
Aborted (core dumped)
```
What gives?

## In summary:

- `libegs_app.so` is needed, but it also reports 'egspp: Failed to resolve the address of the 'createApplication' function in the application library /home/mainegrahe/EGSnrc/egs_home/bin/new-viewer/libegs_app.so'
- viewer works with the old applications on my `$EGS_HOME` for instance from the `develop` branch
- It crashes if the applications are updated with the version in [this branch](https://github.com/Xuan-Zheng05/EGSnrc/tree/add-egs-view-editor3)
- After further testing, even with the old apps the viewer starts crashing after starting it many times on the command line issuing the message `free(): corrupted unsorted chunks`
- when one closes the viewer, it issues the message `corrupted size vs. prev_size`
- After the second restart, closing the viewer keeps the cmd line busy for a while

**System:** gcc version 13.3.0 (Ubuntu 13.3.0-6ubuntu2~24.04.1)

**Comment by rtownson** [comment]
Thanks for the testing Ernesto! I wonder if this is an LD_LIBRARY_PATH problem? As far as I know the general instructions to getting it to work should be:

1. compile egs++
2. compile egs_view
3. copy application from HEN_HOUSE to EGS_HOME
4. compile application
5. `export LD_LIBRARY_PATH=$HEN_HOUSE/egs++/dso/new-viewer;$HEN_HOUSE/pieces/linux/egs_view_64 Dgas-NE2571-HDR.egsinp`

**Comment by rtownson** [comment]
@mainegra  I failed to reproduce any issues with a new install in ubuntu 24, gcc 13. I guess they are related to you being on an existing repo?

I tried having one configuration on develop and one called new-view, but still can't reproduce the issue..

**Comment by ftessier** [comment]
@rtownson, before I review, I would like to confirm that this is our working integration branch for egs_view, on [Xuan-Zheng05](https://github.com/Xuan-Zheng05/EGSnrc/tree/add-egs-view-editor3) I mean. I burnt (myself) in the estart pr! 😃

**Comment by ftessier** [comment]
This is evidently a huge pr, with changes across the entire code base to implement a new viewer feature (input file editing). It looks as though changes are pretty similar across the board however, which is both a positive (repeated patterns) and a negative (loss of specific attention across similar edits).

Here is a helpful review guide by Claude Sonnet 4.6, as a starting point (not necessarily all true, be careful):

## PR #1164 — Review map

### What this PR actually is: three things bundled together

**Thing 1 — The core new abstraction: `egs_input_struct`**
A new compile-time metadata facility where every egs++ component (geometry, source, shape, ausgab object, application) declares its valid input keys, types, allowed values, and inter-key dependencies via a `getInputs()` virtual method returning an `EGS_InputStruct*`. This is the scaffolding that powers the editor.

**Thing 2 — The editor itself: `egs_editor` + `egs_view` changes**
A new Qt tab in egs_view with syntax highlighting, autocompletion driven by Thing 1, input validation (red-line comments), search/replace, block indentation, undo history, example snippets menu, and an application selector dropdown. Also includes dark mode and a splitter layout between controls and image.

**Thing 3 — Compiler/Qt modernization (should arguably be separate PRs)**
Qt6 compatibility, gcc14 fixes, Fortran `-std=legacy`, `register` keyword removal for C++17. These overlap with PRs #1341 and #1342.

---

### The boilerplate map: what is equivalent, what is not

**Practically equivalent — review one exemplar per category, spot-check the rest:**

| Category | Files | Pattern | Risk |
|---|---|---|---|
| Geometry hooks | Every `.cpp` in `egs++/geometry/egs_*/` | `getInputs()` returning static input metadata | Purely documentary — wrong allowed values cause confusing editor behaviour, not simulation bugs. The astyle commit obscures signal here. |
| Source hooks | Every `.cpp` in `egs++/sources/egs_*/` | Same pattern | Same low physics risk; `egs_radionuclide_source` is the most complex (lots of keys, recently touched by #1381) |
| Shape hooks | `egs++/egs_shapes.cpp` etc. | Same pattern | Low risk |
| Ausgab hooks | Every `.cpp` in `egs++/ausgab_objects/egs_*/` | Same pattern | `egs_fluence_scoring` is flagged below as non-equivalent |
| Application hooks | `egs_chamber`, `egs_fac`, `egs_cbct`, `egs_gammaspec` | Same pattern but through a new DSO mechanism | **Non-equivalent — see flags below** |

**Good exemplars to read carefully:**
- Geometry: `egs_sphere` (trivial baseline) and `egs_cylinders` (medium complexity with multiple types)
- Source: `egs_point_source` vs. `egs_radionuclide_source`
- One ausgab object, e.g. `egs_dose_scoring`

---

### Flagged issues requiring scrutiny (in order of priority)

🔴 **`libegs_app.so` — new hard runtime dependency with crash on absence** (reported by @mainegra)

🔴 **`createApplication` symbol not found — a DSO interface mismatch** (reported by @mainegra)

🟠 **Heap corruption on repeated restarts** (reported by @mainegra; commits `76b5889`, `36d7bfc` address symptoms — root cause TBD)

🟠 **Duplicate input block collisions — ad hoc disambiguation strategy** (commits `707e2ba`, `e42d24e`)

🟡 **`getInputs()` virtual method contract — ownership, null-safety, thread safety**

🟡 **`egs_fluence_scoring` axis-label regression** (introduced in `a7ab9c3`, reportedly fixed in `1415fbb` — verify present in current HEAD)

---

### Candidates for extraction to separate PRs

| What | Rationale |
|---|---|
| Qt6 compatibility (`62107ec`, `7bff625`) | Overlaps with #1341; clean and self-contained |
| Compiler/Fortran fixes (`b7fde09`, `2f8a905`, `d84c01d`, `5db2182`, `99a7240`, `44770e6`) | Overlaps with #1342; completely independent of the editor feature |
| Dark mode + splitter layout (`cb53997`, `f348565`) | Pure UI improvements unrelated to the editor or the `egs_input_struct` infrastructure |

The `egs_input_struct` infrastructure + all `getInputs()` hooks + the `egs_editor` Qt widget are mutually dependent and should stay together.

---

### Bottom line

The boilerplate `getInputs()` flood across geometries, sources, shapes and ausgab objects is **low-risk** and can be reviewed efficiently by sampling. The real review surface is: the `egs_app` DSO loading architecture (crashes), the `EGS_InputStruct` ownership/lifecycle design (potential leaks and thread safety), and the input namespace disambiguation strategy (currently ad hoc). Everything else follows from getting those three right.

**Comment by rtownson** [comment]
> @rtownson, before I review, I would like to confirm that this is our working integration branch for egs_view, on [Xuan-Zheng05](https://github.com/Xuan-Zheng05/EGSnrc/tree/add-egs-view-editor3) I mean. I burnt (myself) in the estart pr! 😃

Yes, I have been working directly on Xuan's branch.

**Comment by ftessier** [comment]
I will start by extracting orthogonal commits into new prs. Wait let me rebase on develop first, maybe some of these commits will drop out already.

**Comment by ftessier** [comment]
Rebased on `develop`, diff with original pr branch is minimal (this is obscured if you click on the `Compare` link, on account of everything that was merged into develop in the mean time. Here is the actual diff:

```diff
diff --git a/HEN_HOUSE/egs++/egs_run_control.h b/HEN_HOUSE/egs++/egs_run_control.h
index 16c0ab44..e5102b73 100644
--- a/HEN_HOUSE/egs++/egs_run_control.h
+++ b/HEN_HOUSE/egs++/egs_run_control.h
@@ -53,12 +53,12 @@ class EGS_Input;
 inline void addRunControlBlock(shared_ptr<EGS_InputStruct> blockPtr) {
     shared_ptr<EGS_BlockInput> runBlock = blockPtr->addBlockInput("run control");
     runBlock->addSingleInput("ncase", true, "The number of histories to simulate.");
-    runBlock->addSingleInput("nbatch", false, "The number of batches to divide the simulation into. After each batch, a checkpoint is created to allow for simulation restarts. For simulations with very large intermediate data (like a high resolution dose grid), reducing nbatch to 1 can speed up the simulation. Defaults to 10.");
+    runBlock->addSingleInput("nbatch", false, "The number of batches to divide the simulation into. After each batch, a checkpoint is created to allow resuming the simulation. For simulations with very large intermediate data (like a high resolution dose grid), reducing nbatch to 1 can speed up the simulation. Defaults to 10.");
     runBlock->addSingleInput("nchunk", false, "For parallel runs, each job splits its portion of the simulation into nchunk serial chunks, and each of those chunks is divided into nbatch batches. After each chunk, the parallel job grabs a new portion of the simulation to work on. Defaults to 10.");
     runBlock->addSingleInput("max cpu hours allowed", false, "The number hours after which the simulation will be haulted. Defaults to -1, which is no limit.");
     runBlock->addSingleInput("statistical accuracy sought", false, "The statistical uncertainty for a particular quantity of interest, below which the simulation will be haulted. Note that the quantity must be defined by the application (e.g. the cavity dose in egs_chamber), and in general is undefined (e.g. this input does nothing for egs_app).");
     runBlock->addSingleInput("geometry error limit", false, "The number of geometry errors that will be allowed to occur, before haulting the simulation. Defaults to 0.");
-    runBlock->addSingleInput("calculation", false, "The calculation type: first (default, runs a new simulation), restart (resumes a terminated simulation), analyze (prints results), combine (combines results from a parallel run). Defaults to 'first'.", {"first", "restart", "analyze", "combine"});
+    runBlock->addSingleInput("calculation", false, "The calculation type: first (default, runs a new simulation), resume (continue a crashed simulation), analyze (prints results), combine (combines results from a parallel run). Defaults to 'first'.", {"first", "resume", "analyze", "combine"});
 }
 
 inline string addRunControlExample() {
diff --git a/HEN_HOUSE/egs++/view/viewcontrol.cpp b/HEN_HOUSE/egs++/view/viewcontrol.cpp
index 4c42b0d9..5574fcc8 100644
--- a/HEN_HOUSE/egs++/view/viewcontrol.cpp
+++ b/HEN_HOUSE/egs++/view/viewcontrol.cpp
@@ -1129,7 +1129,7 @@ bool GeometryViewControl::loadInput(bool reloading, EGS_BaseGeometry *simGeom) {
         newGeom = simGeom;
     }
 
-    // restart from scratch (copied from main.cpp)
+    // relaunch from scratch (copied from main.cpp)
     EGS_Float xmin = -50, xmax = 50;
     EGS_Float ymin = -50, ymax = 50;
     EGS_Float zmin = -50, zmax = 50;
@@ -1197,9 +1197,7 @@ bool GeometryViewControl::loadInput(bool reloading, EGS_BaseGeometry *simGeom) {
     RenderParameters &rp = gview->pars;
     rp.allowRegionSelection = allowRegionSelection;
     rp.trackIndices.assign(6,1);
-
-    gview->restartWorker();
-
+    gview->relaunchWorker();
     if (!simGeom) {
         setGeometry(newGeom,user_colors,xmin,xmax,ymin,ymax,zmin,zmax,reloading);
         origSimGeom = g;
@@ -3866,7 +3864,7 @@ void GeometryViewControl::particleSlider(EGS_Float slidertime) {
         int startindex=1,endindex=1;
 
         // loop through the photons time index list one at a time
-        for (int j=0; j<timelist_p.size(); j++) {
+        for (size_t j=0; j<timelist_p.size(); j++) {
 
             // check whether the jth photon time index is within the +/- half
             // time window range centered at the current time index
@@ -3908,7 +3906,7 @@ void GeometryViewControl::particleSlider(EGS_Float slidertime) {
         endindex=1;
 
         // loop through the electrons time index list one at a time
-        for (int j=0; j<timelist_e.size(); j++) {
+        for (size_t j=0; j<timelist_e.size(); j++) {
 
             // check whether the jth electron time index is within the +/- half
             // time window range centered at the current time index
@@ -3948,7 +3946,7 @@ void GeometryViewControl::particleSlider(EGS_Float slidertime) {
         startindex=1;
         endindex=1;
         // loop through the positron time index list one at a time
-        for (int j=0; j<timelist_po.size(); j++) {
+        for (size_t j=0; j<timelist_po.size(); j++) {
             // check whether the jth positron time index is within the +/- half
             // time window range centered at the current time index
             if (timelist_po[j]<(slidertime+(t_window/2)) && timelist_po[j]>(slidertime-(t_window/2))) {
@@ -4441,3 +4439,19 @@ vector<string> findDensityCorrectionInputs(string compound_dir) {
 
     return fileList;
 }
+
+bool GeometryViewControl::hasValidTime() {
+    // This function is used to determine whether the time index elements should be shown in the case that no dynamic geometry is present
+    bool incltime;
+
+    ifstream data(filename_tracks.toUtf8().constData(), ios::binary);
+
+    // Skip the first few bits related to the string head_inctime
+    data.seekg(sizeof(head_inctime));
+    // Read the boolean of whether or not time indices are included
+    data.read((char *)&incltime, sizeof(bool));
+
+    data.close();
+
+    return incltime;
+}
diff --git a/HEN_HOUSE/egs++/view/viewcontrol.h b/HEN_HOUSE/egs++/view/viewcontrol.h
index c6448a7f..0974dea9 100644
--- a/HEN_HOUSE/egs++/view/viewcontrol.h
+++ b/HEN_HOUSE/egs++/view/viewcontrol.h
@@ -250,4 +250,4 @@ protected slots:
 
 };
 
-#endif // GEOMETRYVIEWCONTROL_H
\ No newline at end of file
+#endif // GEOMETRYVIEWCONTROL_H
diff --git a/HEN_HOUSE/user_codes/egs_cbct/egs_cbct.cpp b/HEN_HOUSE/user_codes/egs_cbct/egs_cbct.cpp
index 7aa175a8..3c8dc068 100644
--- a/HEN_HOUSE/user_codes/egs_cbct/egs_cbct.cpp
+++ b/HEN_HOUSE/user_codes/egs_cbct/egs_cbct.cpp
@@ -4323,7 +4323,7 @@ extern "C" {
         outputBlock->setAppName("egs_cbct");
         outputBlock->addSingleInput("display type", false, "Whether to record the total, attenuated (primary), or scatter Kerma. Defaults to total.", {"total", "attenuated", "scattered"});
         outputBlock->addSingleInput("store signal map", false, "Whether to a printout of the detector scan in a .egsmap file.", {"yes", "no"});
-        outputBlock->addSingleInput("store data arrays", false, "Whether to store .egsdat files during the simulation, to allow for restarting crashed runs. Defaults to yes.", {"yes", "no"});
+        outputBlock->addSingleInput("store data arrays", false, "Whether to store .egsdat files during the simulation, to allow for resuming crashed runs. Defaults to yes.", {"yes", "no"});
         outputBlock->addSingleInput("verbose", false, "Whether to print extra information during the simulation. Defaults to no.", {"yes", "no"});
         shared_ptr<EGS_BlockInput> outputScanBlock = outputBlock->addBlockInput("scan output");
         outputScanBlock->addSingleInput("scan file", false, "The filename to output the scan from the simulation.");
```

**Comment by rtownson** [comment]
The extra `hasValidTime` is a mistake, there's now a duplicate of that function.

**Comment by ftessier** [comment]
Good eye @rtownson, I fixed that. Folded it inside the HEAD~1 commit removing duplicated code, I had missed that block! 🙏🏻

**Comment by ftessier** [comment]
Cleaned up the commit title and messages. @rtownson this can be merged into the `develop` branch, but it needs a bit more scrutiny before the release to `master` v2026.

**Comment by ftessier** [comment]
@rtownson, rebasing on the updated develop, can you help me with resolving this conflict? In the end is it `my geometry` or `base geometry`?

```diff
diff --cc HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
index d40cd31b,3f771811..00000000
--- a/HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
+++ b/HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
@@@ -78,8 -78,7 +78,14 @@@ A dynamic geometry is defined usin
  :start geometry:
      name        = ...
      library     = egs_dynamic_geometry
++<<<<<<< HEAD
 +    my geometry = name of a predefined geometry that we want to add motion to
 +    :start motion: # units of cm and degrees
++||||||| parent of 605e889b (Add dynamic geometry inputs to egs_view editor)
++    my geometry = name of a predefined geometry that we want to add motion to :start motion: # units of cm and degrees
++=======
+     base geometry = name of a predefined geometry that we want to add motion to :start motion: # units of cm and degrees
++>>>>>>> 605e889b (Add dynamic geometry inputs to egs_view editor)
          control point = timeIndex(1) xtrans(1) ytrans(1) ztrans(1) xrot(1) yrot(1) zrot(1)
          control point = timeIndex(2) xtrans(2) ytrans(2) ztrans(2) xrot(2) yrot(2) zrot(2)
```

**Comment by ftessier** [comment]
Rebased on the updated `develop` branch, resolving 3 conflicts (plus the deletion of mevegs.cpp):

- [x] Trivial conflict (accept current):

```diff
diff --cc HEN_HOUSE/egs++/egs_base_geometry.cpp
index 060c4680,8e69281c..00000000
--- a/HEN_HOUSE/egs++/egs_base_geometry.cpp
+++ b/HEN_HOUSE/egs++/egs_base_geometry.cpp
@@@ -1196,15 -1174,5 +1196,21 @@@ int EGS_BaseGeometry::setLabels(const s

      return 1;
  }
++<<<<<<< HEAD
 +
 +bool EGS_BaseGeometry::validateRegions(const vector<int> &regions) {
 +
 +    // validate region numbers (stop upon first invalid region)
 +    for (int reg : regions) {
 +        if (reg < 0 || reg >= nreg) {
 +            egsWarning("EGS_BaseGeometry::validateRegions(): geometry %s, region %d is out of bounds\n", getName().c_str(), reg);
 +            return false;
 +        }
 +    }
 +    return true;
 +}
++||||||| parent of 230cde6a (Add a text editor to egs_view)
++=======
+
+
++>>>>>>> 230cde6a (Add a text editor to egs_view)
```

- [x] Accept both changes to contributor list:

```diff
diff --cc HEN_HOUSE/egs++/ausgab_objects/egs_dose_scoring/egs_dose_scoring.cpp
index 728a6637,f4fe8e1d..00000000
--- a/HEN_HOUSE/egs++/ausgab_objects/egs_dose_scoring/egs_dose_scoring.cpp
+++ b/HEN_HOUSE/egs++/ausgab_objects/egs_dose_scoring/egs_dose_scoring.cpp
@@@ -29,10 -29,7 +29,11 @@@
  #                   Blake Walters
  #                   Max Orok
  #                   Martin Martinov
 +#                   Dave Macrillo
 +#                   Matt Ronan
 +#                   Nigel Vezeau
 +#                   Lou Thompson
+ #                   Hannah Gallop
  #
  ###############################################################################
```

- [x] Picked `base geometry` key name (accept incoming):

```diff
diff --cc HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
index d40cd31b,3f771811..00000000
--- a/HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
+++ b/HEN_HOUSE/egs++/geometry/egs_dynamic_geometry/egs_dynamic_geometry.h
@@@ -78,8 -78,7 +78,14 @@@ A dynamic geometry is defined usin
  :start geometry:
      name        = ...
      library     = egs_dynamic_geometry
++<<<<<<< HEAD
 +    my geometry = name of a predefined geometry that we want to add motion to
 +    :start motion: # units of cm and degrees
++||||||| parent of 605e889b (Add dynamic geometry inputs to egs_view editor)
++    my geometry = name of a predefined geometry that we want to add motion to :start motion: # units of cm and degrees
++=======
+     base geometry = name of a predefined geometry that we want to add motion to :start motion: # units of cm and degrees
++>>>>>>> 605e889b (Add dynamic geometry inputs to egs_view editor)
          control point = timeIndex(1) xtrans(1) ytrans(1) ztrans(1) xrot(1) yrot(1) zrot(1)
          control point = timeIndex(2) xtrans(2) ytrans(2) ztrans(2) xrot(2) yrot(2) zrot(2)
```

**Comment by ftessier** [comment]
@rtownson, merging this as is on top of `develop`, but this needs some scrutiny before merging in the v2026 release. There are a few potential memory leaks I would like to review.
